File: homework2/Saytam_Mishra/q3_binarySearchTree.java
package homework2.Saytam_Mishra;

class BinarySearchTree {
    class Node {
        int data;
        Node left;
        Node right;

        public Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    public BinarySearchTree() {
        this.root = null;
    }

    public int min() { // O(logn) in average case and O(n) in worst case
        if(root == null) {
            throw new IllegalStateException("Tree is empty");
        }

        Node current = root;

        while(current.left != null) {
            current = current.left;
        }

        return current.data;
    }

    public int max() { // O(logn) in average case and O(n) in worst case
        if(root == null) {
            throw new IllegalStateException("Tree is empty");
        }

        Node current = root;

        while(current.right != null) {
            current = current.right;
        }

        return current.data;
    }

    public boolean contains(int val) { // O(logn) in average case and O(n) in worst case
        Node current = root;

        while(current != null) {
            if(current.data > val) {
                current = current.left;
            } else if(current.data < val) {
                current = current.right;
            } else {
                return true;
            }
        }

        return false;
    }

    public void insert(int val) { // O(logn) in average case and O(n) in worst case
        Node newNode = new Node(val);

        if(root == null) {
            root = newNode;
            return;
        }

        Node current = root;
        Node parent = null;

        while(current != null) {
            parent = current;
            if(current.data > val) {
                current = current.left;
            } else if(current.data < val) {
                current = current.right;
            } else {
                return; // Node with data val already exists
            }
        }

        if(parent.data > val) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
    }

    public void delete(int val) { // O(logn) in average case and O(n) in worst case
        if(root == null) return;

        if(root.data == val) {
            root = helper(root);
            return;
        }

        Node current = root;

        while (current != null) {
            if(current.data > val) {
                if (current.left != null && current.left.data == val) {
                    current.left = helper(current.left);
                    return;
                } else {
                    current = current.left;
                }
            } else {
                if (current.right != null && current.right.data == val) {
                    current.right = helper(current.right);
                    return;
                } else {
                    current = current.right;
                }
            } 
        }
    }

    public Node helper(Node root) {
        if(root.left == null) {
            return root.right;
        }

        if(root.right == null) {
            return root.left;
        }

        Node rightNode = root.right;

        Node rightMostNode = findRightMostNode(root.left);

        rightMostNode.right = rightNode;
        
        return root.left;
    }

    public Node findRightMostNode(Node root) {
        if(root.right == null) {
            return root;
        }

        return findRightMostNode(root.right);
    }

    public void printInOrder() {
        System.out.print("In-order: ");
        inOrderHelper(this.root);
        System.out.println();
    }

    private void inOrderHelper(Node node) {
        if (node == null) {
            return;
        }

        inOrderHelper(node.left);

        System.out.print(node.data + " ");

        inOrderHelper(node.right);
    }
}

public class q3_binarySearchTree {

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();

        bst.insert(20);
        bst.insert(10);
        bst.insert(30);
        
        bst.printInOrder();
    }
}

// Time taken - 36m 8s

File: homework2/Vedant_Pahariya/Q3_BinarySearchTree.cpp
struct Node {
  int data;
  Node* left;
  Node* right;
};

class BinarySearchTree {
  private:
    Node* root;

  public:
  // Constructor
    BinarySearchTree() {
        root = NULL;
    }

    // returns the minimum value in the BST.  O(logn) time.
    int min(){
        if(root == NULL){
            return 0;
        }

         Node* current = root;
        while(current->left != NULL) {
            current = current->left;
        }
        return current->data;
    }

    // returns the maximum value in the BST.  O(logn) time.
    int max(){
        if(root == NULL){
            return 0;
        }

        Node* current = root;
        while(current->right != NULL) {
            current = current->right;
        }
        return current->data;
    }

    // returns a boolean indicating whether val is present in the BST.  O(logn) time. 
    bool contains(int val){
        Node* current = root;
        while(current != NULL) {
            if(current->data == val) {
                return true;
            } else if(current->data > val) {
                current = current->left;
            } else {
                current = current->right;
            }
        }
        return false;
    }
    
    // creates a new Node with data val in the appropriate location. For simplicity, do not allow duplicates. If val is already present, insert is a no-op. O(logn) time.
    void insert(int val){
        // create new node
        Node* newNode = new Node();
        newNode->data = val;
        newNode->left = NULL;
        newNode->right = NULL;

        //find the point where to insert
        Node* current = root;

        while(current != NULL){
            if(current->data == val){
                return; // no-op for duplicate
            } else if (current->data > val){
                if(current->left == NULL){
                    current->left = newNode;
                    return;
                }
                current = current->left;
            } 
            else {
                if(current->right == NULL){
                    current->right = newNode;
                    return;
                }
                current = current->right;
            }
        }
        root = newNode; 
    }   
    
    // deletes the Node with data val, if it exists. O(logn) time.
    // we can't use delete keyword because it is a reserved keyword in C++
    // deletes the Node with data val, if it exists. O(logn) time. took help of SUing tnr fxn
void deleteNode(int val) {
    if(root == NULL) {
        return;
    }
    
    Node* current = root;
    Node* parent = NULL;
    
    // Find the node to delete and its parent
    while(current != NULL && current->data != val) {
        parent = current;
        if(val < current->data) {
            current = current->left;
        } else {
            current = current->right;
        }
    }
    
    // Node not found
    if(current == NULL) {
        return;
    }
    
    // Case 1: Node has no children (leaf node)
    if(current->left == NULL && current->right == NULL) {
        if(parent == NULL) {
            // Deleting root
            root = NULL;
        } else if(parent->left == current) {
            parent->left = NULL;
        } else {
            parent->right = NULL;
        }
        delete current;
    }
    // Case 2: Node has only right child
    else if(current->left == NULL) {
        if(parent == NULL) {
            // Deleting root
            root = current->right;
        } else if(parent->left == current) {
            parent->left = current->right;
        } else {
            parent->right = current->right;
        }
        delete current;
    }
    // Case 3: Node has only left child
    else if(current->right == NULL) {
        if(parent == NULL) {
            // Deleting root
            root = current->left;
        } else if(parent->left == current) {
            parent->left = current->left;
        } else {
            parent->right = current->left;
        }
        delete current;
    }
    // Case 4: Node has two children
    else {
        // Find inorder successor (smallest node in right subtree)
        Node* successorParent = current;
        Node* successor = current->right;
        
        while(successor->left != NULL) {
            successorParent = successor;
            successor = successor->left;
        }
        
        // Copy successor's data to current node
        current->data = successor->data;
        
        // Delete successor (it has at most one right child)
        if(successorParent->left == successor) {
            successorParent->left = successor->right;
        } else {
            successorParent->right = successor->right;
        }
        delete successor;
    }
}
}

